ext/random/gammasection.c

```diff
@@ -0,0 +1,115 @@
+/*
+   +----------------------------------------------------------------------+
+   | Copyright (c) The PHP Group                                          |
+   +----------------------------------------------------------------------+
+   | This source file is subject to version 3.01 of the PHP license,      |
+   | that is bundled with this package in the file LICENSE, and is        |
+   | available through the world-wide-web at the following url:           |
+   | https://www.php.net/license/3_01.txt                                 |
+   | If you did not receive a copy of the PHP license and are unable to   |
+   | obtain it through the world-wide-web, please send a note to          |
+   | license@php.net so we can mail you a copy immediately.               |
+   +----------------------------------------------------------------------+
+   | Authors: Tim Düsterhus <timwolla@php.net>                            |
+   |                                                                      |
+   | Based on code from: Frédéric Goualard                                |
+   +----------------------------------------------------------------------+
+*/
+
+#ifdef HAVE_CONFIG_H
+# include "config.h"
+#endif
+
+#include "php.h"
+#include "php_random.h"
+#include <math.h>
+
+/* This file implements the γ-section algorithm as published in:
+ *
+ * Drawing Random Floating-Point Numbers from an Interval. Frédéric
+ * Goualard, ACM Trans. Model. Comput. Simul., 32:3, 2022.
+ * https://doi.org/10.1145/3503512
+ */
+
+static double gamma_low(double x)
+{
+	return x - nextafter(x, -DBL_MAX);
+}
+
+static double gamma_high(double x)
+{
+	return nextafter(x, DBL_MAX) - x;
+}
+
+static double gamma_max(double x, double y)
+{
+	return (fabs(x) > fabs(y)) ? gamma_high(x) : gamma_low(y);
+}
+
+static uint64_t ceilint(double a, double b, double g)
+{
+	double s = b / g - a / g;
+	double e;
+
+	if (fabs(a) <= fabs(b)) {
+		e = -a / g - (s - b / g);
+	} else {
+		e = b / g - (s + a / g);
+	}
+
+	double si = ceil(s);
+
+	return (s != si) ? (uint64_t)si : (uint64_t)si + (e > 0);
+}
+
+PHPAPI double php_random_gammasection_closed_open(const php_random_algo *algo, php_random_status *status, double min, double max)
+{
+	double g = gamma_max(min, max);
+	uint64_t hi = ceilint(min, max, g);
+	uint64_t k = 1 + php_random_range64(algo, status, hi - 1); /* [1, hi] */
```

Comment:
I would prefer that these APIs didn't throw, to be honest -- I like it when the C APIs can be used basically like regular C code without Zend Engine side effects getting in the way, but I'm not going to block the PR on it.

Reply:
@morrisonlevi

> or does 0 work out acceptably in the math and then the error will still be set in the global state?

0 is a valid return value that may be returned by php_random_range64 in the happy path, so the result will be well-defined in the error path (even if not particularly random).

> I would prefer that these APIs didn't throw,

Depending on the algo (“the engine”) these functions may call into userland, thus needing to deal with Exceptions is unavoidable:

```
<?php
final class ThrowingEngine implements Random\Engine {
	public function generate(): string
	{
		throw new \Exception('test');
	}
}
$r = new Random\Randomizer(new ThrowingEngine());

$r->getFloat(1, 2);
```

Reply:
To clarify, though: `php_random_range64` _will not_ throw, unless either the underlying algo throws, or the underlying algo is so horribly biased that it may not be called random (a number is rejected with < 50% chance, 50 rejections thus have a probability of < 2<sup>-50</sup> which is less than 1 in a quadrillion).